plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt

```diff
@@ -0,0 +1,162 @@
+import org.jetbrains.org.objectweb.asm.MethodVisitor
+import org.jetbrains.org.objectweb.asm.Opcodes
+import org.jetbrains.org.objectweb.asm.commons.Method
+
+enum class AbiMethodInfo {
+    KEEP,
+    STRIP,
+}
+
+sealed class AbiClassInfo {
+    object Public : AbiClassInfo()
+    class Stripped(val methodInfo: Map<Method, AbiMethodInfo>) : AbiClassInfo()
+}
+
+/**
+ * Record ABI information for all classes in the current compilation unit.
+ *
+ * This needs to be a `ClassBuilderInterceptor`, since we need the descriptors
+ * in order to know which methods can be safely stripped from the output.
+ * On the other hand we cannot produce any output in this pass, since we need
+ * to know which classes are part of the public ABI in order to produce the
+ * correct `InnerClasses` attributes.
+ *
+ * ---
+ *
+ * Classes which are private, local, or anonymous can be stripped unless
+ * they are marked as part of the public ABI by a bit in the Kotlin
+ * metadata annotation. Public ABI classes have to be kept verbatim, since
+ * they are copied to all call sites of a surrounding inline function.
+ *
+ * If we keep a class we will strip the bodies from all non-inline function,
+ * remove private functions, and copy inline functions verbatim. There is one
+ * exception to this for inline suspend functions.
+ *
+ * For an inline suspend function `f` we will usually generate two methods,
+ * a non-inline method `f` and an inline method `f$$forInline`. The former can
+ * be stripped. However, if `f` is not callable directly, we only generate a
+ * single inline method `f` which should be kept.
+ */
+class JvmAbiClassBuilderInterceptor : ClassBuilderInterceptorExtension {
+    val abiClassInfo: MutableMap<String, AbiClassInfo> = mutableMapOf()
+
+    override fun interceptClassBuilderFactory(interceptedFactory: ClassBuilderFactory, bindingContext: BindingContext, diagnostics: DiagnosticSink): ClassBuilderFactory =
+        object : DelegatingClassBuilderFactory(interceptedFactory) {
+            override fun newClassBuilder(origin: JvmDeclarationOrigin): DelegatingClassBuilder =
+                AbiInfoClassBuilder(interceptedFactory.newClassBuilder(origin))
+        }
+
+    private inner class AbiInfoClassBuilder(private val delegate: ClassBuilder) : DelegatingClassBuilder() {
+        lateinit var internalName: String
+        var localOrAnonymousClass = false
+        var publicAbi = false
+        val methodInfos = mutableMapOf<Method, AbiMethodInfo>()
+        val maskedMethods = mutableSetOf<Method>() // Methods which should be stripped even if they are marked as KEEP
+
+        override fun getDelegate(): ClassBuilder = delegate
+
+        override fun defineClass(origin: PsiElement?, version: Int, access: Int, name: String, signature: String?, superName: String, interfaces: Array<out String>) {
+            // Always keep annotation classes
+            if (access and Opcodes.ACC_ANNOTATION != 0) {
```

Comment:
Is there any test for private annotations?

Reply:
To be honest, this was added to fix a problem that occurred in our internal codebase and I don't remember what exactly the issue was.

In general we need to keep annotation classes for Java compatibility, but I don't remember which aspect of an annotation class is in danger of being stripped at all. I'll try to reproduce the failure once I can run said internal codebase with a more up to date version of the Kotlin compiler again.

Meanwhile I've added a test in `plugins/jvm-abi-gen/testData/content/annotation` so that we have at least some test coverage for annotation classes.